src/store/styleguide/typescript-react-analyzer/index.ts

```diff
@@ -0,0 +1,114 @@
+import { getReactComponentExports, ReactComponentExport } from './typescript/react';
+import { PatternFileInfo, ReactPattern } from './react-pattern';
+import { StyleguideAnalyzer } from '../styleguide-analyzer';
+import * as ts from 'typescript';
+
+import * as FileUtils from 'fs';
+import * as PathUtils from 'path';
+import { PatternType } from '../../pattern/pattern-type';
+
+export class TypescriptReactAnalyzer extends StyleguideAnalyzer<ReactPattern> {
+	protected program: ts.Program;
+
+	public getPatternType(): PatternType {
+		return PatternType.react;
+	}
+
+	public analyze(path: string): ReactPattern[] {
+		const patterns: ReactPattern[] = [];
+
+		const fileInfos = walkDirectoriesAndCollectPatterns(path);
+
+		const declarationFiles = fileInfos.map(patternInfo => patternInfo.declarationFilePath);
+
+		this.program = ts.createProgram(declarationFiles, {}, undefined, this.program);
+
+		fileInfos.forEach(fileInfo => {
+			const sourceFile = this.program.getSourceFile(fileInfo.declarationFilePath);
+			const exports = getReactComponentExports(sourceFile, this.program);
+
+			exports.forEach(exportInfo => {
+				const id = generatePatternId(path, fileInfo, exportInfo);
+				const name = getPatternName(fileInfo, exportInfo);
+
+				const pattern = new ReactPattern({
+					id,
+					name,
+					analyzer: this,
+					fileInfo,
+					exportInfo
+				});
+
+				patterns.push(pattern);
+			});
+		});
+
+		return patterns;
+	}
+}
+
+function getPatternName(fileInfo: PatternFileInfo, exportInfo: ReactComponentExport): string {
+	const baseName = PathUtils.basename(fileInfo.jsFilePath, '.js');
+	const directoryName = PathUtils.basename(fileInfo.directory);
+	const name = exportInfo.name ? exportInfo.name : baseName !== 'index' ? baseName : directoryName;
+
+	return name;
+}
+
+function walkDirectoriesAndCollectPatterns(directory: string): PatternFileInfo[] {
```

Comment:
Should this not be moved outside of the `TypescriptReactAnalyzer`?
This function would also be used by every other analyzer or am i missing something?

Reply:
I think we should leave it where it is until we know how pattern detection strategies for other libraries or conventions would work. We also wouldn't know the return type of the function since other conventions require different files to make a valid pattern?

Imagine angular has a template file and a js file that don't neccissarily have to be located in the same directory.

Perhaps you already have a solution in mind?